example/heatEquation/src/heatEquation.cpp

```diff
@@ -0,0 +1,210 @@
+/* Copyright 2019 Benjamin Worpitz, Matthias Werner
+ *
+ * This file exemplifies usage of Alpaka.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED “AS IS” AND ISC DISCLAIMS ALL WARRANTIES WITH
+ * REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL ISC BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
+ * IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <alpaka/alpaka.hpp>
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+
+
+//#############################################################################
+//! Alpaka version of explicit finite-difference 1d heat equation solver
+//!
+//! Solving equation u_t(x, t) = u_xx(x, t) using a simple explicit scheme with
+//! forward difference in t and second-order central difference in x
+//!
+//! \param uCurrBuf grid values of u for each x and the current value of t:
+//!                 u(x, t) | t = t_current
+//! \param uNext resulting grid values of u for each x and the next value of t:
+//!              u(x, t) | t = t_current + dt
+//! \param extent number of grid nodes in x (eq. to numNodesX)
+//! \param dx step in x
+//! \param dt step in t
+
+struct HeatEquationKernel
+{
+    template<
+        typename TAcc
+    >
+    ALPAKA_FN_ACC auto operator()(
+        TAcc const & acc,
+        double const * const uCurrBuf,
+        double * const uNextBuf,
+        double const & extent,
+        double const & dx,
+        double const & dt
+
+    )const -> void
+    {
+        // Each kernel executes one element
+        double const r = dt / ( dx * dx );
+        int idx = alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0];
+        if(idx > 0 && idx < extent){
```

Comment:
IMO this is the BUG: it must be

```
if(idx > 0 && idx < extent - 1u)
{ ...
```

because you access later `uCurrBuf[idx + 1]` and this will be outside of the buffer for `idx=extent -1`